### ISSUE TITLE
BnP Avro input to string/string output

### ISSUE THREAD

**Issue by elad** [voldemort/voldemort#421]
Is it possible to run a BnP job, where the input is Avro, and the output is a read-only store with both the key and the value as simple strings?

It seems to me that once the input is Avro, that also forces the output to be Avro as well. That creates a ton of problems because it seems that client access to the cluster now needs to be Avro-aware, and my use case doesn't need Avro once data is in Voldemort.

**Comment by arunthirupathi** [comment]
Avro uses UTF-8 to serialize Strings. Voldemort String does the same. Not
sure whether there are any gotchas in the Voldemort client, but you should
be able to use StoreClient<String,String> on Avro UTF-8. There could be
some inconsistencies as I haven't tried this.

Thanks,
Arun.

**Comment by elad** [comment]
Take a look at the error I'm getting:

```
Key serializer differs:
	BnP config/data has:	SerializerDefinition(name = avro-generic, schema-info = {0={"type":"string","avro.java.string":"String"}}, compression = null)
	Voldemort server has:	SerializerDefinition(name = string, schema-info = {}, compression = null)
Value serializer differs:
	BnP config/data has:	SerializerDefinition(name = avro-generic, schema-info = {0="bytes"}, compression = null)
	Voldemort server has:	SerializerDefinition(name = bytes, schema-info = {}, compression = null)
```

**Comment by arunthirupathi** [comment]
Store definition should say Avro Generic, (not versioned) and String. But
Voldemort Client can be StoreClient<String,String>

Thanks,
Arun.

**Comment by elad** [comment]
@arunthirupathi I'm not sure what you mean by "Voldemort Client can be StoreClient<String,String>" - there's no Java code involved in this process. Just store definitions on the nodes and BnP configuration. Are you saying that the problem is that the schema-info is specified with a version 0?

**Comment by arunthirupathi** [comment]
Sorry for not being clear.

Avro String and Voldemort String serializes to same byte representation and
may be back. But Build and Push Job enforces strict check on the Schema so
it will not match. But if you have a client that reads from Voldemort (
general way to do that is by creating a StoreClient, you will directly get
a String).

In other words, what is the problem if you have to define the Store as
Key/Value Schema as Avro String ?

Thanks,
Arun.

**Comment by elad** [comment]
@arunthirupathi sorry for the delay - I was on vacation.

The way I see it, Voldemort has a design flaw if the output must be Avro as well once the input is Avro. We use Avro as a format for processing in Hadoop. Once we're done with Hadoop, we need the data accessible in a different format. I don't see why I still have to fuss with Avro when I don't really need it. The problem is that I don't want to add dependencies and code that I don't really need just because the underlying datastore tightly couples input and output format. :/

Voldemort has all the information to do this itself: it knows the input is Avro, it knows which fields it needs to look at for key/value, and it knows the output should be strings, so it can validate those fields are strings as well.

Is it still possible to use a custom serializer? Maybe then I can remove the "Avroness" from the data and keep the format knowledge internally...

**Comment by FelixGV** [comment]
It might be possible to use non-Avro serializers in BnP. Some people did mention doing that on the mailing list, but it is not something we support internally at LinkedIn. We don't have much experience with this kind of set up, so we can't offer much guidance.

It might "happen to already work", or it may require a fairly simple patch to polish some of the rough edges. I don't know... I'm afraid you're on your own on this own. You might want to search the mailing list and reply to some of the old threads or start a new one, in order to get some feedback from the rest of the community.

**Comment by arunthirupathi** [comment]
As per the mailing list discussions, Voldemort 1.6.9 used to support String
serialization. But 1.10.* only supports Avro or JSON.

As Felix pointed out, Voldemort RO is only tested always with Avro and JSON.

PS : Community had success with using Voldemort RO setup with zoned stores
as well, but that is not intended as well and 1.10.* errors out on these
conditions.

Thanks,
Arun.

**Comment by elad** [comment]
Fair enough. When the time comes, I'll either write my own "coordinator" that knows how to deal with the Avro store and/or custom serializers to avoid Avro knowledge in the resulting stores.

@arunthirupathi why was this functionality removed in 1.10.*, btw?
